api/src/main/java/org/openmrs/module/pihcore/apploader/ConfigAppLoaderFactory.java

```diff
@@ -0,0 +1,135 @@
+                        if (f.getName().endsWith(suffix)) {
+                            files.add(f);
+                        }
+                    }
+                }
+            }
+        }
+        return files;
+    }
+
+    @Override
+    public List<AppTemplate> getAppTemplates() {
+        List<AppTemplate> templates = new ArrayList<>();
+        for (File f : getConfigFilesBySuffix("AppTemplates.json")) {
+            try {
+                List<AppTemplate> l = objectMapper.readValue(f, new TypeReference<List<AppTemplate>>() {});
+                templates.addAll(l);
+            }
+            catch (Exception e) {
+                logger.error("Error reading AppTemplates configuration file: {}", f.getName(), e);
+            }
+        }
+        return templates;
+    }
+
+    @Override
+    public List<AppDescriptor> getAppDescriptors() {
+        List<AppDescriptor> descriptors = new ArrayList<>();
+        for (File f : getConfigFilesBySuffix("app.json")) {
+            try {
+                List<AppDescriptor> apps = objectMapper.readValue(f, new TypeReference<List<AppDescriptor>>() {});
+                for (int i=0; i<apps.size(); i++) {
+                    AppDescriptor app = apps.get(i);
+                    // Allow setting the id from the file path, to avoid having to set this directly when not needed
+                    if (StringUtils.isBlank(app.getId())) {
+                        app.setId(getFileNameAsId(f) + (apps.size() > 1 ? "_" + i : ""));
+                    }
+                    if (app.getExtensions() != null) {
+                        for (int y=0; y<app.getExtensions().size(); y++) {
+                            Extension extension = app.getExtensions().get(y);
+                            // Allow setting the extension id from the app id
+                            if (StringUtils.isBlank(extension.getId())) {
+                                extension.setId(app.getId() + (app.getExtensions().size() > 1 ? "_" + y : ""));
+                            }
+                            // Allow setting the appId of the extension automatically from the containing app id
+                            if (StringUtils.isBlank(extension.getAppId())) {
+                                extension.setAppId(app.getId());
+                            }
+                            else {
+                                if (!extension.getAppId().equals(app.getId())) {
+                                    logger.error("Extension has appId {} that does not match app id {}", extension.getAppId(), app.getId());
+                                }
+                            }
+                        }
+                    }
+                    descriptors.add(app);
+                }
+            }
+            catch (Exception e) {
+                logger.error("Error reading AppDescriptors configuration file: {}", f.getName(), e);
+            }
+        }
+        return descriptors;
+    }
+
+    @Override
```

Comment:
I don't expect we will actually use this, but if we need to have free-standing extensions, this will load an file under `configuration/appframework/*/*extension.json`